clouddriver-google-common/src/main/groovy/com/netflix/spinnaker/clouddriver/googlecommon/batch/GoogleBatchRequest.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2018 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.googlecommon.batch;
+
+import com.google.api.client.googleapis.batch.BatchRequest;
+import com.google.api.client.googleapis.batch.json.JsonBatchCallback;
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpRequestInitializer;
+import com.google.api.services.compute.Compute;
+import com.google.api.services.compute.ComputeRequest;
+import com.google.common.collect.Lists;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Helper class for sending batch requests to GCE.
+ */
+@Slf4j
+public class GoogleBatchRequest {
+
+  private static final int MAX_BATCH_SIZE = 100; // Platform specified max to not overwhelm batch backends.
+  private static final int DEFAULT_CONNECT_TIMEOUT_MILLIS = (int) TimeUnit.MINUTES.toMillis(2);
+  private static final int DEFAULT_READ_TIMEOUT_MILLIS = (int) TimeUnit.MINUTES.toMillis(2);
+
+  private List<QueuedRequest> queuedRequests;
+  private String clouddriverUserAgentApplicationName;
+  private Compute compute;
+
+  public GoogleBatchRequest(Compute compute, String clouddriverUserAgentApplicationName) {
+    this.compute = compute;
+    this.clouddriverUserAgentApplicationName = clouddriverUserAgentApplicationName;
+    this.queuedRequests = new ArrayList<>();
+  }
+
+  public void execute() {
+    if (queuedRequests.size() == 0) {
+      log.debug("No requests queued in batch, exiting.");
+      return;
+    }
+
+    List<BatchRequest> queuedBatches = new ArrayList<>();
+    List<List<QueuedRequest>> requestPartitions = Lists.partition(queuedRequests, MAX_BATCH_SIZE);
+    requestPartitions.forEach(requestPart -> {
+      BatchRequest newBatch = newBatch();
+      requestPart.forEach(qr -> {
+        try {
+          qr.getRequest().queue(newBatch, qr.getCallback());
+        } catch (IOException ioe) {
+          log.error("Queueing request {} in batch failed.", qr);
+          throw new RuntimeException(ioe);
+        }
+      });
+      queuedBatches.add(newBatch);
+    });
+
+    ExecutorService threadPool = new ForkJoinPool(10);
```

Comment:
This is from the prior PR, but is there a reason to create a new pool each time? Would it be more efficient to have a common pool (or use the default pool) in this case?

Reply:
We could but it's unclear what the performance implications would be for piping all the batches through one pool since we can't tell the size of the batches beforehand when deciding the pool parallelism. We _do_ have an understanding of rough sizes per batch, and which batch sizes are considered large. I'm inclined to keep this as is and tune it later if there are problems.